gk/fib.c

```diff
@@ -0,0 +1,1409 @@
+
+		ret = setup_neighbor_tbl(socket_id, (identifier * 2 + 1),
+			ETHER_TYPE_IPv6, gk_conf->max_num_ipv6_neighbors,
+			&neigh_fib_ipv6->u.neigh6);
+		if (ret < 0)
+			goto init_fib_ipv6;
+
+		if (iface == &net_conf->front)
+			neigh_fib_ipv6->action = GK_FWD_NEIGHBOR_FRONT_NET;
+		else if (likely(iface == &net_conf->back))
+			neigh_fib_ipv6->action = GK_FWD_NEIGHBOR_BACK_NET;
+		else
+			rte_panic("Unexpected condition at %s: invalid interface %s!\n",
+				__func__, iface->name);
+
+		ret = gk_lpm_add_ipv6_route(iface->ip6_addr.s6_addr,
+			iface->ip6_addr_plen, fib_id, ltbl);
+		if (ret < 0)
+			goto free_fib_ipv6_ht;
+
+		*neigh6_fib = neigh_fib_ipv6;
+	}
+
+	return 0;
+
+free_fib_ipv6_ht:
+	destroy_neigh_hash_table(&neigh_fib_ipv6->u.neigh);
+
+init_fib_ipv6:
+	initialize_fib_entry(neigh_fib_ipv6);
+
+free_fib_ipv4:
+	if (neigh_fib_ipv4 == NULL)
+		return -1;
+
+	*neigh_fib = NULL;
+
+	rte_lpm_delete(ltbl->lpm, ntohl(iface->ip4_addr.s_addr),
+		iface->ip4_addr_plen);
+
+free_fib_ipv4_ht:
+	destroy_neigh_hash_table(&neigh_fib_ipv4->u.neigh);
+
+init_fib_ipv4:
+	initialize_fib_entry(neigh_fib_ipv4);
+
+out:
+	return -1;
+}
+
+static int
+init_fib_tbl(struct gk_config *gk_conf)
+{
+	int ret;
+	unsigned int i;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+	struct gk_fib *neigh_fib_front = NULL, *neigh6_fib_front = NULL;
+	struct gk_fib *neigh_fib_back = NULL, *neigh6_fib_back = NULL;
+
+	rte_spinlock_init(&ltbl->lock);
+
+	for (i = 0; i < gk_conf->gk_max_num_ipv4_fib_entries; i++) {
+		ltbl->fib_tbl[i].action = GK_FIB_MAX;
+		rte_atomic16_init(&ltbl->fib_tbl[i].num_updated_instances);
+	}
+
+	for (i = 0; i < gk_conf->gk_max_num_ipv6_fib_entries; i++) {
+		ltbl->fib_tbl6[i].action = GK_FIB_MAX;
+		rte_atomic16_init(&ltbl->fib_tbl6[i].num_updated_instances);
+	}
+
+	/* Set up the FIB entry for the front network prefixes. */
+	ret = setup_net_prefix_fib(0, &neigh_fib_front,
+		&neigh6_fib_front, &gk_conf->net->front, gk_conf);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to setup the FIB entry for the front network prefixes at %s\n",
+			__func__);
+		goto out;
+	}
+
+	/* Set up the FIB entry for the back network prefixes. */
+	RTE_VERIFY(gk_conf->net->back_iface_enabled);
+	ret = setup_net_prefix_fib(1, &neigh_fib_back,
+		&neigh6_fib_back, &gk_conf->net->back, gk_conf);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to setup the FIB entry for the back network prefixes at %s\n",
+			__func__);
+		goto free_front_fibs;
+	}
+
+	return 0;
+
+free_front_fibs:
+	if (neigh_fib_front != NULL) {
+		struct gatekeeper_if *iface = &gk_conf->net->front;
+		rte_lpm_delete(gk_conf->lpm_tbl.lpm,
+			ntohl(iface->ip4_addr.s_addr), iface->ip4_addr_plen);
+		destroy_neigh_hash_table(&neigh_fib_front->u.neigh);
+		initialize_fib_entry(neigh_fib_front);
+		neigh_fib_front = NULL;
+	}
+	if (neigh6_fib_front != NULL) {
+		struct gatekeeper_if *iface = &gk_conf->net->front;
+		rte_lpm6_delete(gk_conf->lpm_tbl.lpm6,
```

Comment:
It's better to wrap `rte_lpm6_delete()` with `RTE_VERIFY()`: `RTE_VERIFY(rte_lpm6_delete(...) == 0);`